lib/Wikia/src/CircuitBreaker/Storage/APCUAdapter.php

```diff
@@ -0,0 +1,166 @@
+
+	private function readFailure( $key ) {
+		return "APCu fetch failed for key: {$key}";
+	}
+
+	private function storeFailure( $key ) {
+		return "APCu failed storing value for key: {$key}";
+	}
+
+	/**
+	 * @param Configuration $configuration
+	 * @return void
+	 */
+	public function setConfiguration( Configuration $configuration ) {
+		$this->configuration = $configuration;
+	}
+
+	/**
+	 * @param string $service
+	 * @return int
+	 */
+	public function load( $service ) {
+		return (int)apcu_fetch( $service );
+	}
+
+	/**
+	 * @param string $service
+	 * @param int $count
+	 * @return void
+	 */
+	public function save( $service, $count ) {
+		if (apcu_store( $service, $count ) === false) {
+			$this->throwException( $this->storeFailure( $service ) );
+		}
+	}
+
+	/**
+	 * @param string $service
+	 * @return void
+	 */
+	public function increment( $service ) {
+		$status = false;
+		apcu_inc( $service, 1, $status );
+		if ( $status === false ) {
+			$this->throwException( $this->storeFailure( $service ) );
+		}
+	}
+
+	/**
+	 * decrement failure count
+	 *
+	 * If the operation would decrease the value below 0, the new value must be 0.
+	 *
+	 * @param string $service
+	 * @return void
+	 */
+	public function decrement( $service ) {
+		$status = false;
+		apcu_dec( $service, 1, $status );
+		if ( $status === false ) {
+			$this->throwException( $this->storeFailure( $service ) );
+		}
+	}
+
+	/**
+	 * sets last failure time
+	 *
+	 * @param string $service
+	 * @param int $lastFailureTime
+	 * @return void
+	 */
+	public function saveLastFailureTime( $service, $lastFailureTime ) {
+		if (apcu_store( $service, $lastFailureTime ) === false) {
+			$this->throwException( $this->storeFailure( $service ) );
+		}
+	}
+
+	/**
+	 * returns last failure time
+	 *
+	 * @param $service
+	 * @return int | null
+	 */
+	public function loadLastFailureTime( $service ) {
+		$status = false;
+		$value = (int)apcu_fetch( $service, $status );
+		if ( $status ) {
+			return $value;
+		}
+
+		$this->throwException( $this->readFailure( $service ) );
+	}
+
+	/**
+	 * sets status
+	 *
+	 * @param string $service
+	 * @param int $status
+	 * @return void
+	 */
+	public function saveStatus( $service, $status ) {
+		if (apcu_store( $service, $status ) === false) {
+			$this->throwException( $this->storeFailure( $service ) );
+		}
+	}
+
+	/**
+	 * returns status
+	 *
+	 * @param string $service
+	 * @return int
+	 */
+	public function loadStatus( $service ) {
+		$status = apcu_fetch( $service );
+		if ( $status === false && !apcu_exists( $service ) ) {
+			$this->saveStatus( $service, Ganesha::STATUS_CALMED_DOWN );
+
+			return Ganesha::STATUS_CALMED_DOWN;
+		}
+
+		return $status;
+	}
+
+	/**
+	 * resets all counts
+	 *
+	 * @return void
+	 */
+	public function reset() {
+		// according to documentation apcu_clear_cache "Returns TRUE always", so I don't think we
+		// can do much more in terms of handling errors in this case
+		apcu_clear_cache();
```

Comment:
This may cause some unexpected side effects. We are clearing the whole cache and not only keys we care about. Sadly there seems to be no other way to do that correctly.

Reply:
Can we use https://www.php.net/manual/en/apcuiterator.construct.php to safely delete only entries we need?